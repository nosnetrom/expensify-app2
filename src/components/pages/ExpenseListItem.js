import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const amountProcessor = (fullAmt) => {
  let extension = '';
  const amt = parseInt(fullAmt) / 100;
  if (fullAmt % 100 == 0) {
    extension = '.00';
  } else if (fullAmt % 10 == 0) {
    extension = '0';
  }
  return `${amt}${extension}`;
};

const friendlyDate = (d) => {
  const fd = moment(d).format('DD MMM YYYY');
  return fd;
};

// Export a stateless functional component
//  - description, amount, createdAt
const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <tr className="expenseListItemWrapper">
    <td className="itemDescription">
      <Link to={`/edit/${id}`}>{description}</Link>
    </td>
    <td className="itemAmount">{amountProcessor(amount)}</td>
    <td className="itemCreatedAt">{friendlyDate(createdAt)}</td>
  </tr>
);

export default ExpenseListItem;
