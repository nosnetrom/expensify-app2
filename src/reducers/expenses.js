const expensesReducerDefaultState = [];

// Expenses reducer
export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      //console.log(action.expense);
      return [...state, action.expense];
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          // spread the existing expense object, then overwrite with the object as passed in
          return {
            ...expense,
            ...action.updates,
          };
        } else {
          return expense;
        }
      });
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    default:
      return state;
  }
};
