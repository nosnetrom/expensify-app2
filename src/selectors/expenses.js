import moment from 'moment';

// Redux filtering
export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, 'day')
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, 'day')
        : true;
      // const textMatch =
      //   expense.description.toLowerCase().indexOf(text.toLowerCase()) > -1;
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());
      // If all 3 consts == true, then we have a listable expense

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      // Use ternary function for sorting items by order
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};
