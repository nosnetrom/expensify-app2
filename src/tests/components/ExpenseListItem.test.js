import React from 'react';
import { shallow } from 'enzyme';
//import toJSON from 'enzyme-to-json';
import expenses from '../fixtures/expenses';
import ExpenseListItem from '../../components/pages/ExpenseListItem';

test('should render ExpenseListItem selected from expenses', () => {
  const item = expenses[0];
  const wrapper = shallow(<ExpenseListItem {...item} />);
  expect(wrapper).toMatchSnapshot();
});
