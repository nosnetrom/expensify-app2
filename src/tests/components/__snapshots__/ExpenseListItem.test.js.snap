// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem selected from expenses 1`] = `
<tr
  className="expenseListItemWrapper"
>
  <td
    className="itemDescription"
  >
    <Link
      to="/edit/1"
    >
      USB stick
    </Link>
  </td>
  <td
    className="itemAmount"
  >
    8.99
  </td>
  <td
    className="itemCreatedAt"
  >
    12 Jan 1970
  </td>
</tr>
`;
